idea/idea-frontend-independent/src/org/jetbrains/kotlin/idea/quickfix/ReplaceCallFix.kt

```diff
@@ -0,0 +1,116 @@
+        val element = element ?: return
+        val elvis = element.elvisOrEmpty(notNullNeeded)
+        val betweenReceiverAndOperation = element.elementsBetweenReceiverAndOperation().joinToString(separator = "") { it.text }
+        val newExpression = KtPsiFactory(element).createExpressionByPattern(
+            "$0$betweenReceiverAndOperation$operation$1$elvis",
+            element.receiverExpression,
+            element.selectorExpression!!
+        )
+        val replacement = element.replace(newExpression)
+        if (elvis.isNotEmpty()) {
+            replacement.moveCaretToEnd(editor, project)
+        }
+    }
+
+    private fun KtQualifiedExpression.elementsBetweenReceiverAndOperation(): List<PsiElement> {
+        val receiver = receiverExpression
+        val operation = operationTokenNode as? PsiElement ?: return emptyList()
+        val start = receiver.nextSibling?.takeIf { it != operation } ?: return emptyList()
+        val end = operation.prevSibling?.takeIf { it != receiver } ?: return emptyList()
+        return PsiTreeUtil.getElementsOfRange(start, end)
+    }
+}
+
+class ReplaceImplicitReceiverCallFix(
+    expression: KtExpression,
+    private val notNullNeeded: Boolean
+) : KotlinPsiOnlyQuickFixAction<KtExpression>(expression) {
+    override fun getFamilyName() = text
+
+    override fun getText() = KotlinBundle.message("replace.with.safe.this.call")
+
+    override fun invoke(project: Project, editor: Editor?, file: KtFile) {
+        val element = element ?: return
+        val elvis = element.elvisOrEmpty(notNullNeeded)
+        val newExpression = KtPsiFactory(element).createExpressionByPattern("this?.$0$elvis", element)
+        val replacement = element.replace(newExpression)
+        if (elvis.isNotEmpty()) {
+            replacement.moveCaretToEnd(editor, project)
+        }
+    }
+}
+
+class ReplaceWithSafeCallFix(
+    expression: KtDotQualifiedExpression,
+    notNullNeeded: Boolean
+) : ReplaceCallFix(expression, "?.", notNullNeeded) {
+    override fun getText() = KotlinBundle.message("replace.with.safe.call")
+}
+
+class ReplaceWithSafeCallForScopeFunctionFix(
+    expression: KtDotQualifiedExpression,
+    notNullNeeded: Boolean
+) : ReplaceCallFix(expression, "?.", notNullNeeded) {
+    override fun getText() = KotlinBundle.message("replace.scope.function.with.safe.call")
+}
+
+class ReplaceWithDotCallFix(expression: KtSafeQualifiedExpression) : ReplaceCallFix(expression, "."), CleanupFix {
+    override fun getText() = KotlinBundle.message("replace.with.dot.call")
+
+    companion object : KotlinSingleIntentionActionFactory() {
+        override fun createAction(diagnostic: Diagnostic): IntentionAction? {
+            val qualifiedExpression = diagnostic.psiElement.getParentOfType<KtSafeQualifiedExpression>(strict = false) ?: return null
+            return ReplaceWithDotCallFix(qualifiedExpression)
+        }
+    }
+}
+
+fun KtExpression.elvisOrEmpty(notNullNeeded: Boolean): String {
```

Comment:
I was about to suggest make these two `private`, but due to the other usage I mentioned below, it's better to place them back to `Utils`.